### PR TITLE
dm-relay/streamer: move binlog log reader into relay

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve? 
it's first part of subtask 4 in [#2214 of DM](https://github.com/pingcap/dm/issues/2214)

no code logic change in this pr, to make review/merge easier 

### What is changed and how it works?
- move binlog log reader into relay
- move `binlog/writer` to relay
- flat small sub packages
- rename some structs to avoid conflict

### Check List 

no code logic change

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by Ehco1996 — APPROVED** [review]
LGTM

**Comment by D3Hunter** [comment]
@lance6716 PTAL

**Comment by D3Hunter** [comment]
@glorv @lichunzhu PTAL

**Review by lance6716 — COMMENTED** [review]
will review after some more urgent PRs

File: dm/relay/remote_reader.go [hunk 1]
```diff
@@ -74,8 +74,8 @@ type reader struct {
 }
 
 // NewReader creates a Reader instance.
-func NewReader(cfg *Config) Reader {
-	return &reader{
+func NewReader(cfg *RConfig) Reader {
```
Comment (lance6716) [inline dm/relay/remote_reader.go:77]: now we have relay.NewBinlogReader and relay.NewReader, need a better name.

Or in futre we can wrap relay.NewBinlogReader into another public method like relay.StartPulling?

**Review by lance6716 — COMMENTED** [review]
rest lgtm

there're 2 cases that I can't decide.

- we plan to move remote reader into relay package, so the retry logic of remote reader is also moved. 
   Now retry logic is put in a separated file, that's quite small so I prefer we include it into remote_reader.go. (too many files under relay folder IMO 😂 )
- we plan to move relay log purger into relay package, so the strategy of purger is also moved. 
   The strategies are located in 4 files and they're not relevant to other files (except for purger), so I'd like to merge them into `purger_strategy.go`

File: dm/relay/remote_reader.go [hunk 2]
```diff
@@ -11,7 +11,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package reader
+package relay
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply dm/relay/remote_reader.go:14]: remote_reader -> upstream_reader to follow current naming convention, `_reader` is kept to make it easier to know what's inside

i hate they both have `_reader` suffix too, but remove it makes more ambiguities to me.

File: dm/relay/remote_reader.go (hunk 1 above)
Comment (D3Hunter) [reply dm/relay/remote_reader.go:77]: will be that way in next pr.

**Comment by D3Hunter** [comment]
> too many files under relay folder IMO

not a problem to me as long as we can difference between them

**Review by lance6716 — APPROVED** [review]

**Pull request merged**